sympy/logic/algorithms/lra_theory.py

```diff
@@ -0,0 +1,667 @@
+                    conflict.add(Boundry(xi, self.upper[xi][0], True, self.up_origin[xi], self.upper[xi][1] != 0))
+
+                    conflict = {-self.boundry_rev_enc[c] for c in conflict}
+                    return False, conflict
+                xj = sorted(cand, key=lambda v: str(v))[0]
+                _debug_internal_state_printer2(xi, xj)
+                M = self._pivot_and_update(M, basic, nonbasic, xi, xj, self.upper[xi])
+
+    def _pivot_and_update(self, M, basic, nonbasic, xi, xj, v):
+        i, j = basic[xi], self.col_index[xj]
+        assert M[i, j] != 0
+        theta_lhs = (v[0] - self.assign[xi][0])/M[i, j]
+        theta_rhs = (v[1] - self.assign[xi][1])/M[i, j]
+        self.assign[xi] = v
+        self.assign[xj] = (self.assign[xj][0] + theta_lhs, self.assign[xj][1] + theta_rhs)
+        for xk in basic:
+            if xk != xi:
+                k = basic[xk]
+                akj = M[k, j]
+                self.assign[xk] = (self.assign[xk][0] + akj*theta_lhs, self.assign[xk][1] + akj*theta_rhs)
+        # pivot
+        basic[xj] = basic[xi]
+        del basic[xi]
+        nonbasic.add(xi)
+        nonbasic.remove(xj)
+        return self._pivot(M, i, j)
+
+    @staticmethod
+    def _pivot(M, i, j):
+        """
+        Performs a pivot operation about entry i, j of M by performing
+        a series of row operations on a copy of M and returing the result.
+        The original M is left unmodified.
+
+        Conceptually, M represents a system of equations and pivoting
+        can be thought of as rearranging equation i to be in terms of
+        variable j and then substiututing in the rest of the equations
+        to get rid of other occurances of variable j.
+
+        Example
+        =======
+
+        >>> from sympy.matrices.dense import Matrix
+        >>> from sympy.logic.algorithms.lra_theory import LRASolver
+        >>> from sympy import var
+        >>> Matrix(3, 3, var('a:i'))
+        Matrix([
+        [a, b, c],
+        [d, e, f],
+        [g, h, i]])
+
+        This matrix is equivalent to:
+        0 = a*x + b*y + c*z
+        0 = d*x + e*y + f*z
+        0 = g*x + h*y + i*z
+
+        >>> LRASolver._pivot(_, 1, 0)
+        Matrix([
+        [ 0, -a*e/d + b, -a*f/d + c],
+        [-1,       -e/d,       -f/d],
+        [ 0,  h - e*g/d,  i - f*g/d]])
+
+        We rearage equation 1 in terms of variable 0 (x)
+        and substitute to remove x from the other equations.
+
+        0 = 0 + (-a*e/d + b)*y + (-a*f/d + c)*z
+        0 = -x + (-e/d)*y + (-f/d)*z
+        0 = 0 + (h - e*g/d)*y + (i - f*g/d)*z
+        """
+        _, _, Mij = M[i, :], M[:, j], M[i, j]
+        if Mij == 0:
+            raise ZeroDivisionError("Tried to pivot about zero-valued entry.")
+        A = M.copy()
+        A[i, :] = -A[i, :]/Mij
+        for row in range(M.shape[0]):
+            if row != i:
+                A[row, :] = A[row, :] + A[row, j] * A[i, :]
+
+        return A
```

Comment:
@oscarbenjamin could you give any advice on how to speed up this pivot function? Currently it's the bottleneck in the code.

Reply:
<img width="589" alt="Screen Shot 2023-08-18 at 7 06 29 PM" src="https://github.com/sympy/sympy/assets/67246777/81b35817-2eb9-4c98-89b8-2722cf732cb8">

Reply:
How did you reach to the example?
I would take a look at if there is a regression about assignment because of the introduction of `RepMatrix`
https://github.com/sympy/sympy/pull/21626

Reply:
> How did you reach to the example?
I made 20 largish random problems using the code in test_lra_theory.py and used the LRASolver to solve them. 

> I would take a look at if there is a regression about assignment because of the introduction of RepMatrix
https://github.com/sympy/sympy/pull/21626

I'm a bit confused about what this means. Regression as in a test for some fixed bug? What do you mean by "assignment"?

Reply:
Assignment here means `M[i, j] = ...`

Reply:
> I made 20 largish random problems using the code in test_lra_theory.py and used the LRASolver to solve them.

Can you refer to the specific test case?

Reply:
```
problems = [make_random_problem(num_variables=8, num_constraints=16, rational=False, disable_strict=False,
                                  disable_nonstrict=False, disable_equality=False) for _ in range(20)]

problems = [prob_to_enc(cons) for cons in problems]

from pyinstrument import Profiler

profiler = Profiler()
profiler.start()
for enc in problems:
    lra, x_subs, s_subs = LRASolver.from_encoded_cnf(enc)
    assert_bounds(lra, enc)
    lra.check()

profiler.stop()
#profiler.open_in_browser()
```

The screenshot is from lra.check().

Since `num_variables=8, num_constraints=16`, the matrices involved would be around 16 by 24. Basically with these settings each random problem has 16 equalities, strict or non-strict inequalities and each of those constraints would have 8 variables. And the numbers in the matrices would be small integers at first.

Reply:
Anyway, I can confirm that there is no much difference about forcing the Matrix to `EXRAW`.
It means that we may need hard time around, to optimize the algorithm itself.

Reply:
`Matrix.copyin_matrix` can be made more efficient by adding a corresponding DomainMatrix method and using that.